linera-views/src/backends/lru_caching.rs

```diff
@@ -50,54 +58,267 @@ static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
 /// keep track of this.
 
 /// The data structures
+#[allow(clippy::type_complexity)]
 struct LruPrefixCache {
-    map: BTreeMap<Vec<u8>, Option<Vec<u8>>>,
-    queue: LinkedHashMap<Vec<u8>, (), RandomState>,
+    map_find_key_values: BTreeMap<Vec<u8>, Vec<(Vec<u8>, Vec<u8>)>>,
+    map_find_keys: BTreeMap<Vec<u8>, Vec<Vec<u8>>>,
+    map_contains_key: BTreeMap<Vec<u8>, bool>,
+    map_read_value: BTreeMap<Vec<u8>, Option<Vec<u8>>>,
+    queue: LinkedHashMap<(Vec<u8>, CacheEntry), (), RandomState>,
     max_cache_size: usize,
 }
 
 impl<'a> LruPrefixCache {
     /// Creates a LruPrefixCache.
     pub fn new(max_cache_size: usize) -> Self {
         Self {
-            map: BTreeMap::new(),
+            map_find_key_values: BTreeMap::new(),
+            map_find_keys: BTreeMap::new(),
+            map_contains_key: BTreeMap::new(),
+            map_read_value: BTreeMap::new(),
             queue: LinkedHashMap::new(),
             max_cache_size,
         }
     }
 
-    /// Inserts an entry into the cache.
-    pub fn insert(&mut self, key: Vec<u8>, value: Option<Vec<u8>>) {
-        match self.map.entry(key.clone()) {
+    /// Clearing entries of the queue if too long
+    fn insert_queue(&mut self, full_key: (Vec<u8>, CacheEntry)) {
+        self.queue.insert(full_key, ());
+        if self.queue.len() > self.max_cache_size {
+            let Some(value) = self.queue.pop_front() else {
+                unreachable!()
+            };
+            match value.0 {
+                (v, CacheEntry::FindKeyValues) => {
+                    self.map_find_key_values.remove(&v);
+                }
+                (v, CacheEntry::FindKeys) => {
+                    self.map_find_keys.remove(&v);
+                }
+                (v, CacheEntry::ContainsKey) => {
+                    self.map_contains_key.remove(&v);
+                }
+                (v, CacheEntry::ReadValue) => {
+                    self.map_read_value.remove(&v);
+                }
+            }
+        }
+    }
+
+    /// Inserts a read_value entry into the cache.
+    pub fn insert_read_value(&mut self, key: Vec<u8>, value: Option<Vec<u8>>) {
+        let lower_bound = {
+            let iter = self.map_find_keys.iter();
+            let mut suffix_closed_set = SuffixClosedMapIterator::new(iter);
+            let lower_bound = suffix_closed_set.find_prefix(&key);
+            lower_bound.map(|(key_prefix, _)| key_prefix.to_vec())
+        };
+        if let Some(key_prefix) = lower_bound {
+            self.map_find_keys.remove(&key_prefix);
+            let full_prefix = (key_prefix, CacheEntry::FindKeys);
+            self.queue.remove(&full_prefix);
```

Comment:
Why do we have to remove this entry? If we _read_ a value in that range, it doesn't make the result invalid, does it?

Reply:
This has been changed.

The thing is that it was an update to the LRU cache after the write to the cache.